lib/StackAnalysis/StackAnalysis.cpp

```diff
@@ -383,6 +433,811 @@ void commitToModel(GeneratedCodeBasicInfo &GCBI,
   revng_check(TheBinary.verify(true));
 }
 
+static std::optional<uint64_t> electFSO(const auto &MaybeReturns) {
+  auto It = std::min_element(MaybeReturns.begin(),
+                             MaybeReturns.end(),
+                             [](const auto &LHS, const auto &RHS) {
+                               return LHS.second < RHS.second;
+                             });
+  if (It == MaybeReturns.end())
+    return {};
+  return It->second;
+}
+
+static UpcastablePointer<model::FunctionEdge>
+makeEdge(MetaAddress Destination, model::FunctionEdgeType::Values Type) {
+  model::FunctionEdge *Result = nullptr;
+  if (model::FunctionEdgeType::isCall(Type))
+    Result = new model::CallEdge(Destination, Type);
+  else
+    Result = new model::FunctionEdge(Destination, Type);
+  return UpcastablePointer<model::FunctionEdge>(Result);
+};
+
+template<class FunctionOracle>
+SortedVector<model::BasicBlock>
+CFEPAnalyzer<FunctionOracle>::collectDirectCFG(llvm::BasicBlock *Entry,
+                                               llvm::Function *F) {
+  using namespace llvm;
+  SortedVector<model::BasicBlock> CFG;
+
+  for (auto &BB : *F) {
+    if (GCBI::isJumpTarget(&BB)) {
+      auto Start = getBasicBlockPC(&BB);
+      MetaAddress End;
+      uint64_t Size;
+
+      bool Stop = false;
+      model::BasicBlock Block{ Start };
+      OnceQueue<BasicBlock *> Queue;
+      Queue.insert(&BB);
+
+      while (!Queue.empty()) {
+        BasicBlock *Current = Queue.pop();
+
+        auto [CandidateEnd, CandidateSize] = getPC(Current->getTerminator());
+        if (CandidateEnd > End) {
```

Comment:
`End` is invalid on first iteration

Reply:
Right, and then? Invalid as in zero, thus `if (CandidateEnd > End)` is entered, no? What's wrong, sorry?

Reply:
Are you sure `Invalid` always compares lower than anything else? It wasn't (explicitly) designed for that.
If you want to introduce this semantic, we can `static_assert` it in `MetaAddress.h`.
Otherwise, just add an explicit validity check here.

Reply:
Please review if this looks good.